### PR TITLE
Use pagination in upload letters task

### PR DESCRIPTION
**Pull request by banderous**
### Change description ###
Hibernate's streams must be called from within a transaction, which is undesirable for batch processing tasks since the transaction will be rolled back if there is a problem anywhere in the job, including for entities that were processed successfully.

This change processes letters in batches of 10.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java [hunk 1]
```diff
@@ -41,7 +41,6 @@ public UploadLettersTask(
         this.availabilityChecker = availabilityChecker;
     }
 
-    @Transactional
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java]: Wouldn't this be easier if we had moved this to `uploadLetter` and make that method public?

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java [hunk 2]
```diff
@@ -123,4 +126,23 @@ public void should_fail_to_upload_to_sftp_and_stop_from_uploading_any_other_lett
...
+            repository,
+            FtpHelper.getSuccessfulClient(LocalSftpServer.port),
+            availabilityChecker
+        );
+
+        try (LocalSftpServer server = LocalSftpServer.create()) {
+            task.run();
+        }
+        entityManager.clear();
+        assertThat(repository.findByStatus(LetterStatus.Uploaded).count()).isEqualTo(20);
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java]: is 20 the same as BATCH_SIZE * 2?

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java [hunk 3]
```diff
@@ -123,4 +126,23 @@ public void should_fail_to_upload_to_sftp_and_stop_from_uploading_any_other_lett
...
+        UploadLettersTask task = new UploadLettersTask(
+            repository,
+            FtpHelper.getSuccessfulClient(LocalSftpServer.port),
+            availabilityChecker
+        );
+
+        try (LocalSftpServer server = LocalSftpServer.create()) {
+            task.run();
+        }
+        entityManager.clear();
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java]: may be worth adding `// Clear the JPA cache to force a read.` here as well. (or perhaps extract to a method?)

File: src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java (hunk 1 above)
Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java]: If all the letters are marked Uploaded in the same transaction then they either all be committed or all rolled back; having a second transaction won't change this.

Each letter needs to be a separate transaction which is why streams don't work, with streams the order of events is as follows:

BEGIN TRANSACTION
for each letter in stream {
  upload letter
  UPDATE LETTER to Uploaded
}
COMMIT
Nothing will be committed until the iteration completes and the outer transaction commits, and if anything goes wrong the transaction is rolled back.

I'm fairly sure the only way around this is to break this up into multiple transactions. See the [last comment here](https://jira.spring.io/browse/DATAJPA-1023)

Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java]: Though we might be talking about different things; I still need to fix the cause of the SQLException in sandbox, that requires a separate investigation.

Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTask.java]: Still, the streams approach is not going to work if we have 100,000 letters to upload.

File: src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java (hunk 2 above)
Comment (banderous) [reply src/integrationTest/java/uk/gov/hmcts/reform/sendletter/tasks/UploadLettersTaskTest.java]: good spot

File: src/main/java/uk/gov/hmcts/reform/sendletter/entity/LetterRepository.java [hunk 4]
```diff
@@ -9,6 +11,8 @@
 
 public interface LetterRepository extends JpaRepository<Letter, UUID> {
 
+    Page<Letter> findByStatus(LetterStatus status, Pageable pageable);
+
     Stream<Letter> findByStatus(LetterStatus status);
```
[the remark this answers is not in the document]
Comment (banderous) [reply src/main/java/uk/gov/hmcts/reform/sendletter/entity/LetterRepository.java]: Yes but only by tests. I tried getting rid of it but it makes for quite awkward test code.

**Pull request merged**